From 3f3fde84d9dd7517ccd347112089ab11cf56b159 Mon Sep 17 00:00:00 2001
From: Brandon Black <bblack@wikimedia.org>
Date: Wed, 3 Sep 2025 09:07:32 -0500
Subject: [PATCH 087/133] Return runtime err for EISCONN when connecting
X-Developer-Signature: v=1; a=openpgp-sha256; l=2527; i=xtex@astrafall.org;
 h=from:subject; bh=GkC2Y2bJCGwE4c6pYE51IO94FRWrGET3z/idjehcE0A=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78jiUX/XsKz1U72LeXpS9bWGl6poTbb7LmyZ/+st+R
 eLA2ZMNHRNZGMS4GCzFFFmKDBu8WXXS+UWXlcvCzGFlAhkiLdLAwMDAwMLAl5uYV2qkY6Rnqm2o
 Z2ioY6RjxMDFKQBTHdLFyDCLr9e85OO/T3/+f7B3eHniivz1s51+H7iZUt9unTflST4bI8O7zKg
 7nt5SWQ3OczlOHeHIeGDYJrgrMnulC8ffrbyK03kA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

Fixes #25063

(cherry picked from commit 048f9126b8f2b41f9ef103025bafc2773a62d82a)
---
 lib/std/posix.zig | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/lib/std/posix.zig b/lib/std/posix.zig
index 3c7d52f773a5..818065fb9a54 100644
--- a/lib/std/posix.zig
+++ b/lib/std/posix.zig
@@ -4283,6 +4283,9 @@ pub const ConnectError = error{
 
     /// Socket is non-blocking and already has a pending connection in progress.
     ConnectionPending,
+
+    /// Socket was already connected
+    AlreadyConnected,
 } || UnexpectedError;
 
 /// Initiate a connection on a socket.
@@ -4303,7 +4306,7 @@ pub fn connect(sock: socket_t, sock_addr: *const sockaddr, len: socklen_t) Conne
             => return error.NetworkUnreachable,
             .WSAEFAULT => unreachable,
             .WSAEINVAL => unreachable,
-            .WSAEISCONN => unreachable,
+            .WSAEISCONN => return error.AlreadyConnected,
             .WSAENOTSOCK => unreachable,
             .WSAEWOULDBLOCK => return error.WouldBlock,
             .WSAEACCES => unreachable,
@@ -4329,7 +4332,7 @@ pub fn connect(sock: socket_t, sock_addr: *const sockaddr, len: socklen_t) Conne
             .CONNRESET => return error.ConnectionResetByPeer,
             .FAULT => unreachable, // The socket structure address is outside the user's address space.
             .INTR => continue,
-            .ISCONN => unreachable, // The socket is already connected.
+            .ISCONN => return error.AlreadyConnected, // The socket is already connected.
             .HOSTUNREACH => return error.NetworkUnreachable,
             .NETUNREACH => return error.NetworkUnreachable,
             .NOTSOCK => unreachable, // The file descriptor sockfd does not refer to a socket.
@@ -4389,7 +4392,7 @@ pub fn getsockoptError(sockfd: fd_t) ConnectError!void {
             .BADF => unreachable, // sockfd is not a valid open file descriptor.
             .CONNREFUSED => return error.ConnectionRefused,
             .FAULT => unreachable, // The socket structure address is outside the user's address space.
-            .ISCONN => unreachable, // The socket is already connected.
+            .ISCONN => return error.AlreadyConnected, // The socket is already connected.
             .HOSTUNREACH => return error.NetworkUnreachable,
             .NETUNREACH => return error.NetworkUnreachable,
             .NOTSOCK => unreachable, // The file descriptor sockfd does not refer to a socket.
-- 
2.52.0

